python/cudf_polars/cudf_polars/dsl/translate.py

```diff
@@ -25,6 +25,28 @@
 __all__ = ["translate_ir", "translate_named_expr"]
 
 
+class Translator:
+    def __init__(self, visitor: NodeTraverser, *, debug_mode: bool = False):
+        self.visitor = visitor
+        self.debug_mode = debug_mode
+        self.errors = []
+
+
+def debug(func):
+    def wrapper(*args, **kwargs):
+        try:
+            return func(*args, **kwargs)
+        except Exception as e:
+            for arg in args:
+                if isinstance(arg, Translator):
+                    if arg.debug_mode:
+                        arg.errors.append(e)
+                        return ir.ErrorNode(arg.visitor.get_schema(), e)
+            raise
+
+    return wrapper
```

Comment:
*thought/suggestion*: I think now we have our own `Translator` object, we can get away without the decorator.

Idea: the top-level callbacks are `translate_ir` and `translate_expr`, if we leave the existing `singledispatch` functions as-is (no decorator), then we can write something like:


```python
class Translator:
   def __init__(self, visitor, ...):
       self.visitor = visitor
       ...

   def translate_ir(self, *, n: int): # this replaces the free-function translate_ir
       # code from translate_ir function, but when we call _translate_ir we catch errors
       ...
       try:
           return _translate_ir(node, self, schema)
       except Exception as e:
           self.errors.append(e)
           return ir.ErrorNode(schema, e)

   def translate_expr(self, ...): 
       # We can do the same trick here, because translation of an expression can
       # also raise
       ...
       try:
           return _translate_expr(...)
       except Exception as e:
           self.errors.append(e)
           return expr.ErrorNode(dtype, e)
```

Now, afterwards in the callback wrapper, I think we could then do:


```
translator = Translator(visitor)
ir = translator.translate_ir(...)
if len(translator.errors) > 0:
    # print/raise
    raise
set_udf(...)
```